fe/fe-core/src/main/java/com/starrocks/qe/Coordinator.java

```diff
@@ -913,6 +915,47 @@ private void computeFragmentExecParams() throws Exception {
 
     }
 
+    // compute degree of parallelism(dop) for each OlapScanNode in a fragment instance.
+    // the maxScanOperators is numRows of ScanNode divided by session variable pipelineMaxRowCountPerScanOperator,
+    // then each fragment instance take the share of itself own numScanRanges divided by totalNumScanRanges, and
+    // the dop of OlapScanNode is at least 1 and at most minimum(itself own numScanRanges, half the avg number of be cores).
+    private void computeScanDop() {
+        // only take effect in pipeline engine
+        if (ConnectContext.get() == null || !ConnectContext.get().getSessionVariable().isEnablePipelineEngine()) {
+            return;
+        }
+        long maxRowCountPerScanOperator = ConnectContext.get().getSessionVariable().getPipelineMaxRowCountPerScanOperator();
+        int dop = ConnectContext.get().getSessionVariable().getDegreeOfParallelism();
+        for (FragmentExecParams params : fragmentExecParamsMap.values()) {
+            List<PlanNode> scanNodes = params.fragment.getOlapScanNodes(params.fragment.getPlanRoot());
+            for (PlanNode node : scanNodes) {
+                OlapScanNode scanNode = (OlapScanNode) node;
+                int totalNumScanRanges = scanNode.getScanRangeLocations(0).size();
+                long numRows = scanNode.getCardinality();
+                if (scanNode.hasLimit()) {
+                    numRows = Math.min(scanNode.getLimit(), numRows);
+                }
+                int maxNumScanOperators = Math.max(1, (int) (numRows / maxRowCountPerScanOperator));
```

Comment:
maxNumScanOperators not used.

Why not consider input bytes here?

Reply:
these code is unused.
the evaluated row count of OlapScanNode is in-exact, so here just use numScanRanges and degreeOfParallelism to compute dop of ScanNode.